### ISSUE TITLE
[$250] [HOLD for payment 2024-02-20] Implement Client-side Violations when updating money requests

### ISSUE THREAD

**Issue by cead22** [Expensify/App#31093]
- [Design Doc](https://docs.google.com/document/d/1iGvpCoQ6Wt-i2wcpvaMKsWZRNs-oAsWoaS_nLue7_E0/edit)
- The following should only apply to users in the violations beta
- Let's connect EditRequestPage to the POLICY, and TRANSACTION_VIOLATIONS Onyx collections so we can get
    - `requiresTag`, `requiresCategory`, `hasMultipleTagLists`, and `isTaxTrackingEnabled` from the policy, and
    - The transaction violations from Onyx
- In `src/pages/EditRequestPage.js` let's pass policyID to editMoneyRequest, and in there pass it to IOU.updateDistanceRequest and IOU.editMoneyRequest
- Let's update `src/libs/actions/IOU.js `
    - In editMoneyRequest above where [we set optimisticData](https://github.com/Expensify/App/blob/907e65e074d1fec41b2038eacdd78455d2fdb161/src/libs/actions/IOU.js#L1319), call `ViolationUtils.getViolationsOnyxData(transactionChanges, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories)`, and add the return value to the optimisticData
    - Similarly in updateDistanceRequest after [the last call to optimisticData.push](https://github.com/Expensify/App/blob/907e65e074d1fec41b2038eacdd78455d2fdb161/src/libs/actions/IOU.js#L748), we call `ViolationUtils.getViolationsOnyxData(transactionChanges, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories)` and push the return value to optimisticData
- We can set the value of failure data to an object like in the snippet below, where transactionViolations has the data from Onyx
- Once the request is submitted, violations will be calculated on the server in EditMoneyRequest API command and pushed down to the client, but nothing about the violations that are calculated on the client should change

```js
{
    onyxMethod: Onyx.METHOD.MERGE,
    key: `${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`,
    value: transactionViolations,
}
```
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019149143c50ac6f4d</li>
        <li>Upwork Job ID: 1811196753636588997</li>
        <li>Last Price Increase: 2024-07-11</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @mkhutornyi</details>

**Comment by cead22** [comment]
Off hold

**Comment by cead22** [comment]
@lindboe @cdanwards @trevor-coleman can you please comment here so I can assign you?

**Comment by cead22** [comment]
In final review / testing

**Comment by cead22** [comment]
Same ^

**Comment by cead22** [comment]
We prioritized client side violations when **creating** money request. This one comes next

**Comment by lindboe** [comment]
Currently working on this

**Comment by lindboe** [comment]
@cead22 Can you clarify what you mean by "We can set the value of failure data to an object like in the snippet below, where transactionViolations has the data from Onyx"?

I know there's been some discussion about whether to use `Onyx.connect`; does this mean we should also pass in `transactionViolations` from the component? Or is this an acceptable use-case for `Onyx.connect?

**Comment by cead22** [comment]
WIP but waiting on another PR to be merged

**Comment by cdanwards** [comment]
Currently in review

**Comment by cead22** [comment]
@mananjadhav @mkhutornyi sorry for the auto assignment, I'm trying to trigger some _other_ automation

**Comment by cead22** [comment]
@sonialiap can you please handle payment for @parasharrajat for the PR review, and undo this https://github.com/Expensify/App/issues/31093#issuecomment-2221767906, I'm guessing by closing the upwork job? Thanks

**Comment by sonialiap** [comment]
Payment summary 
- @parasharrajat $500 - please request in ND

Upwork job closed

**Comment by JmillsExpensify** [comment]
$500 approved for @parasharrajat
